File: SortingAlgs.py
import time

class SortingAlgs:
    def swap(data, index1, index2):
        temp = data[index1]
        data[index1] = data[index2]
        data[index2] = temp
    
    #bubble sort
    def bubbleSort(data):
        position = len(data) - 1
        while position >= 0:
            scan = 0
            while scan <= position -1:
                if data[scan] > data[scan+1]: #compare to
                    SortingAlgs.swap(data, scan, (scan + 1))
                scan+=1
            position-=1
    
    #quick sort
    def quickSort(data):
        SortingAlgs.quickSortMain(data, 0, len(data) - 1)

    def quickSortMain(data, min, max):
        if min < max:
            indexOfPartition = SortingAlgs.partition(data, min, max)
            SortingAlgs.quickSortMain(data, min, indexOfPartition-1)
            SortingAlgs.quickSortMain(data, indexOfPartition + 1, max)
    
    def partition(data, min, max):
        middle = (min+max)//2
        partitionElement = data[middle]
        SortingAlgs.swap(data, middle, min)
        left = min
        right = max
        while left < right:
            while left < right and data[left]<=(partitionElement): #compare to
                left+=1
            while data[right] > partitionElement:
                right -=1
            if left < right:
                SortingAlgs.swap(data, left, right)
        SortingAlgs.swap(data, min, right)
        return right
    
    #merge sort
    def mergeSort(data):
        SortingAlgs.mergeSortMain(data, 0, len(data)-1)

    def mergeSortMain(data, min, max):
        if min < max:
            mid = (min + max) // 2
            SortingAlgs.mergeSortMain(data, min, mid)
            SortingAlgs.mergeSortMain(data, mid + 1, max)
            SortingAlgs.merge(data, min, mid, max)

    def merge(data, first, mid, last):
        temp = [None]*len(data)
        first1 = first
        last1 = mid
        first2 = mid+1
        last2 = last
        index = first1
        while first1 <= last1 and first2 <= last2:
            if data[first1] < data[first2]:
                temp[index] = data[first1]
                first1 +=1
            else:
                temp[index] = data[first2]
                first2+=1
            index+=1
        while first1 <= last1:
            temp[index] = data[first1]
            first1+=1
            index+=1
        while first2 <= last2:
            temp[index] = data[first2]
            first2+=1
            index+=1
        index = first
        while index <= last:
            data[index] = temp[index]
            index+=1

#testing the algorithms out

start = time.time_ns()
array = [72, 27, 62, 32, 63, 100, 6, 27, 12, 51, 88, 48, 35, 31, 3, 98, 94, 10, 9, 31, 49, 44, 84, 93, 4, 78, 30, 92, 53, 17, 12, 31, 92, 68, 2, 66, 99, 3, 15, 5, 41, 17, 29, 73, 16, 23, 54, 51, 18, 16, 93, 54, 68, 97, 78, 51, 8, 46, 13, 39, 97, 100, 79, 2, 13, 62, 29, 85, 88, 46, 23, 14, 78, 82, 58, 74, 75, 77, 19, 100, 49, 40, 31, 87, 48, 24, 88, 27, 11, 2, 87, 84, 52, 51, 52, 3, 38, 24, 0, 0]
SortingAlgs.bubbleSort(array)
end = time.time_ns()
elapsedTime = (end - start) 
print("Bubble sort time: ",elapsedTime)
print(array)

start2 = time.time_ns()
array2 =[72, 27, 62, 32, 63, 100, 6, 27, 12, 51, 88, 48, 35, 31, 3, 98, 94, 10, 9, 31, 49, 44, 84, 93, 4, 78, 30, 92, 53, 17, 12, 31, 92, 68, 2, 66, 99, 3, 15, 5, 41, 17, 29, 73, 16, 23, 54, 51, 18, 16, 93, 54, 68, 97, 78, 51, 8, 46, 13, 39, 97, 100, 79, 2, 13, 62, 29, 85, 88, 46, 23, 14, 78, 82, 58, 74, 75, 77, 19, 100, 49, 40, 31, 87, 48, 24, 88, 27, 11, 2, 87, 84, 52, 51, 52, 3, 38, 24, 0, 0]
SortingAlgs.quickSort(array2)
end2 = time.time_ns()
elapsedTime2 = (end2-start2)
print("Quick sort time: ", elapsedTime2)
print(array2)

start3=time.time_ns()
array3 = [72, 27, 62, 32, 63, 100, 6, 27, 12, 51, 88, 48, 35, 31, 3, 98, 94, 10, 9, 31, 49, 44, 84, 93, 4, 78, 30, 92, 53, 17, 12, 31, 92, 68, 2, 66, 99, 3, 15, 5, 41, 17, 29, 73, 16, 23, 54, 51, 18, 16, 93, 54, 68, 97, 78, 51, 8, 46, 13, 39, 97, 100, 79, 2, 13, 62, 29, 85, 88, 46, 23, 14, 78, 82, 58, 74, 75, 77, 19, 100, 49, 40, 31, 87, 48, 24, 88, 27, 11, 2, 87, 84, 52, 51, 52, 3, 38, 24, 0, 0]
SortingAlgs.mergeSort(array3)
end3 = time.time_ns()
print("Merge sort time: " + str(end3-start3))
print(array3)

File: SortingAlgs.java
import java.util.*;
import java.lang.*;
import java.io.*;

public class SortingAlgs
{
    private static void swap(int[] data, int index1, int index2)
    {
        int temp = data[index1];
        data[index1] = data[index2];
        data[index2] = temp;
    }

    //bubble sort
    public static void bubbleSort(int[] data)
    {
        int position, scan;
        for (position = data.length - 1; position >= 0; position--)
        {
            for (scan = 0; scan <= position - 1; scan++)
            {
                if (data[scan] > data[scan+1])
                    swap(data, scan, scan + 1);
            }
        }
    }  

    //quick sort
    public static void quickSort(int[] data)
    {
        quickSort(data, 0, data.length - 1);
    }
    
    private static void quickSort(int[] data, int min, int max)
    {
        if (min < max)
        {
            int indexOfPartition = partition(data, min, max); //creates partitions 
            quickSort(data, min, indexOfPartition-1); //sort the left partition 
            quickSort(data, indexOfPartition + 1, max); //sort the right partition 
        }
    }
    
    private static int partition(int[] data, int min, int max)
    {
        int partitionElement;
        int left, right;
        int middle = (min + max)/2;
        partitionElement = data[middle]; //middle data value is partition
        swap(data, middle, min); //moves partition out of the way
        left = min;
        right = max;
        while (left < right)
        {
            //searches for an element that is greater than partition 
            while (left < right && data[left] <= partitionElement)
                left++;
            //searches for an element that is less than partition
            while (data[right] > partitionElement)
                right--;
            //swaps the elements
            if (left < right)
                swap(data, left, right);
        }
        swap(data, min, right); //moves the partition element back to place
        return right;
    }  

    //merge sort 
    public static void mergeSort(int[] data)
    {
        mergeSort(data, 0, data.length-1);
    }
    
    private static void mergeSort(int[] data, int min, int max)
    {
        if (min < max) //stops the recursion when there is only one value in the array
        {
            int mid = (min + max) / 2;
            //splits the array into two 
            mergeSort(data, min, mid);
            mergeSort(data, mid + 1, max);
            merge(data, min, mid, max);
        }
    }

    private static void merge(int[] data, int first, int mid, int last)
    {
        int[] temp = new int[data.length]; //creates a temporary array to put items in
        int first1 = first, last1 = mid; //endpoints of first subarray
        int first2 = mid+1, last2 = last; //endpoints of second subarray
        int index = first1; //index while putting values in the temporary array
        //Copies smaller item of the two arrays into the temporary array until one of the subarrays is used up
        while (first1 <= last1 && first2 <= last2)
        {
            if (data[first1] < data[first2])
            {
                temp[index] = data[first1];
                first1++;
            }
            else
            {
                temp[index] = data[first2];
                first2++;
            }
            index++;
        }
        //copies remaining elements from first subarray, if any
        while (first1 <= last1)
        {
            temp[index] = data[first1];
            first1++;
            index++;
        }
        while (first2 <= last2)
        {
            temp[index] = data[first2];
            first2++;
            index++;
        }
        //copies merged data into original array
        for (index = first; index <= last; index++)
            data[index] = temp[index];
    }

    //for driver method
    public static int[] randArray(int length){
        Random rand = new Random();
        int[] arr = new int[length];
        for (int x = 0; x < length; x++){
            arr[x]=rand.nextInt(length+1);
        }
        return arr;
    }

    public static int mean(ArrayList<Long> arr){
        int sum = 0;
        for (int i=0; i < arr.size(); i++){
            sum += arr.get(i);
        }
        return (sum/arr.size());
    }

    //driver method
    public static void main(String args[])
    {
        ArrayList<Long> bubbleResult = new ArrayList<Long>();
        ArrayList<Long> quickResult = new ArrayList<Long>();
        ArrayList<Long> mergeResult = new ArrayList<Long>();
        int numTrials = 100;
        int arrayLength = 31622;

        for (int trial = 0; trial < numTrials; trial++){
            int[] arr = randArray(arrayLength);
            long start = System.nanoTime();
            int[] newArraylist = arr.clone();
            SortingAlgs.bubbleSort(newArraylist);
            long end = System.nanoTime();
            bubbleResult.add(end-start);

            long start2 = System.nanoTime();
            int[] newArraylist2 = arr.clone();
            quickSort(newArraylist2);
            long end2 = System.nanoTime();
            quickResult.add(end2-start2);

            long start3 = System.nanoTime();
            int[] newArraylist3 = arr.clone();
            mergeSort(newArraylist3);
            long end3 = System.nanoTime();
            mergeResult.add(end3-start3);
            
        }

        int result = mean(bubbleResult);
        System.out.println("Elapsed time for bubble sort: " + result);
        int result1 = mean(quickResult);
        System.out.println("Elapsed time for quick sort: " + result1);
        int result2 = mean(mergeResult);
        System.out.println("Elapsed time for merge sort: " + result2);
    }
}